lib/ansible/parsing/vault/__init__.py

```diff
@@ -526,32 +526,59 @@ def write_data(self, data, filename, shred=True):
 
         if filename == '-':
             sys.stdout.write(b_file_data)
-        else:
-            if os.path.isfile(filename):
-                if shred:
-                    self._shred_file(filename)
-                else:
-                    os.remove(filename)
-            with open(filename, "wb") as fh:
-                fh.write(b_file_data)
+            return
+
+        target_filename, _ = self._unlink_file(filename, shred=shred)
```

Comment:
encrypt_file() and decrypt_file() call write_data(data, filename) with filename being a non-tmp file.